### PR TITLE
trafficfilterapi: Add CRUD and association APIs

### PR DESCRIPTION
**Pull request by marclop**
## Description

Adds an implementation for the traffic filtering APIs so they can be
leveraged in the terraform provider

## Related Issues

Depends on #203 

## Types of Changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: pkg/api/deploymentapi/trafficfilterapi/create.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+// Validate ensures the parameters are usable by Create.
+func (params CreateParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter create params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.Req == nil {
+		merr = merr.Append(errors.New("request payload cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create.go:47]: Optional nit:
```suggestion
		merr = merr.Append(errors.New("request payload is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/create_association.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+// Validate ensures the parameters are usable by CreateAssociation.
+func (params CreateAssociationParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter association create params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("rule set id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create_association.go:53]: optional nit:
```suggestion
		merr = merr.Append(errors.New("rule set id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/create_association.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	if params.EntityID == "" {
+		merr = merr.Append(errors.New("entity id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create_association.go:57]: optional nit:
```suggestion
		merr = merr.Append(errors.New("entity id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/create_association.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("entity type cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create_association.go:61]: optional nit:
```suggestion
		merr = merr.Append(errors.New("entity type is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/delete.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+// Validate ensures the parameters are usable by Delete.
+func (params DeleteParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter delete params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("rule set id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete.go:49]: optional nit:
```suggestion
		merr = merr.Append(errors.New("rule set id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/delete_association.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+// Validate ensures the parameters are usable by DeleteAssociation.
+func (params DeleteAssociationParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter association delete params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("rule set id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete_association.go:52]: optional nit
```suggestion
		merr = merr.Append(errors.New("rule set id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/delete_association.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	if params.EntityID == "" {
+		merr = merr.Append(errors.New("entity id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete_association.go:56]: optional nit:
```suggestion
		merr = merr.Append(errors.New("entity id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/delete_association.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("entity type cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete_association.go:60]: optional nit:
```suggestion
		merr = merr.Append(errors.New("entity type is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/delete_association_test.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		args args
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter association delete params",
+				apierror.ErrMissingAPI,
+				errors.New("rule set id cannot be empty"),
+				errors.New("entity id cannot be empty"),
+				errors.New("entity type cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete_association_test.go:47]: optional nit:
```suggestion
				errors.New("rule set id is not specified and is required for the operation"),
				errors.New("entity id is not specified and is required for the operation"),
				errors.New("entity type is not specified and is required for the operation"),
```

File: pkg/api/deploymentapi/trafficfilterapi/delete_test.go [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+	tests := []struct {
+		name string
+		args args
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter delete params",
+				apierror.ErrMissingAPI,
+				errors.New("rule set id cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/delete_test.go:46]: optional nit:
```suggestion
				errors.New("rule set id is not specified and is required for the operation"),
```

File: pkg/api/deploymentapi/trafficfilterapi/get.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+// Validate ensures the parameters are usable by Get.
+func (params GetParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter get params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("rule set id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/get.go:50]:
```suggestion
		merr = merr.Append(errors.New("rule set id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/get_test.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+		name string
+		args args
+		want *models.TrafficFilterRulesetInfo
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter get params",
+				apierror.ErrMissingAPI,
+				errors.New("rule set id cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/get_test.go:49]: optional nit:
```suggestion
				errors.New("rule set id is not specified and is required for the operation"),
```

File: pkg/api/deploymentapi/trafficfilterapi/update.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+// Validate ensures the parameters are usable by Update.
+func (params UpdateParams) Validate() error {
+	var merr = multierror.NewPrefixed("invalid traffic filter update params")
+	if params.API == nil {
+		merr = merr.Append(apierror.ErrMissingAPI)
+	}
+
+	if params.ID == "" {
+		merr = merr.Append(errors.New("rule set id cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/update.go:50]: optional nit:
```suggestion
		merr = merr.Append(errors.New("rule set id is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/update.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	if params.Req == nil {
+		merr = merr.Append(errors.New("request payload cannot be empty"))
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/update.go:54]: optional nit:
```suggestion
		merr = merr.Append(errors.New("request payload is not specified and is required for the operation"))
```

File: pkg/api/deploymentapi/trafficfilterapi/update_test.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+		args args
+		want *models.TrafficFilterRulesetResponse
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter update params",
+				apierror.ErrMissingAPI,
+				errors.New("rule set id cannot be empty"),
+				errors.New("request payload cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/update_test.go:49]: opitonal nit:
```suggestion
				errors.New("rule set id is not specified and is required for the operation"),
				errors.New("request payload is not specified and is required for the operation"),
```

File: pkg/api/deploymentapi/trafficfilterapi/create_association_test.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+		args args
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter association create params",
+				apierror.ErrMissingAPI,
+				errors.New("rule set id cannot be empty"),
+				errors.New("entity id cannot be empty"),
+				errors.New("entity type cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create_association_test.go:47]:
```suggestion
				errors.New("rule set id is not specified and is required for the operation"),
				errors.New("entity id is not specified and is required for the operation"),
				errors.New("entity type is not specified and is required for the operation"),
```

File: pkg/api/deploymentapi/trafficfilterapi/create_test.go [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+		name string
+		args args
+		want *models.TrafficFilterRulesetResponse
+		err  error
+	}{
+		{
+			name: "fails due to parameter validation",
+			err: multierror.NewPrefixed("invalid traffic filter create params",
+				apierror.ErrMissingAPI,
+				errors.New("request payload cannot be empty"),
```
Comment (karencfv) [inline pkg/api/deploymentapi/trafficfilterapi/create_test.go:48]:
```suggestion
				errors.New("request payload is not specified and is required for the operation"),
```

**Review by karencfv — APPROVED** [review]
Thank you for opening up this PR! 🙇‍♀️ 

I've left a few optional nits on the error messages. The linter is failing but I'm assuming that's because the apidocs update repo hasn't been merged yet? 

I'll leave it as approved so you can update the sdk dependency once that's merged :)

**Comment by marclop** [comment]
That's right Karen it was failing because of the PR it was depending on, thanks for approving it. All the nits should now be addressed.

**Pull request merged**